src/Ouroboros-Mini_Protocols-Chain_Sync.thy

```diff
@@ -0,0 +1,178 @@
+    AwaitReply \<Rightarrow> Idle \<comment> \<open>only for this initial implementation\<close>
+  )"
+
+definition state_machine where
+  [simp]: "state_machine = \<lparr>
+    initial_state = Idle,
+    agent_in_state = agent_in_state',
+    can_finish_in_state = can_finish_in_state',
+    next_state = next_state'
+  \<rparr>"
+
+sublocale chain_sync \<subseteq> protocol_state_machine \<open>state_machine\<close> .
+
+subsection \<open>Programs\<close>
+
+datatype phase =
+  is_intersection_finding: IntersectionFinding |
+  is_items_update: ItemsUpdate
+
+definition roll_back_to :: "'i list \<Rightarrow> 'p \<Rightarrow> ('i \<Rightarrow> 'p) \<Rightarrow> 'i list" where
+  [simp]: "roll_back_to \<I> p \<psi> = (THE \<I>\<^sub>1. \<exists>\<I>\<^sub>2. \<I> = \<I>\<^sub>1 @ \<I>\<^sub>2 \<and> \<psi> (last \<I>\<^sub>1) = p)"
+
+corec client_program where
+  "client_program ph \<kappa> \<psi> \<I> = (
+    case ph of
+      IntersectionFinding \<Rightarrow>
+        \<up> Cont (FindIntersect (\<kappa> \<I>));
+        \<down> M; (partial_case M of
+          Cont IntersectNotFound \<Rightarrow>
+            \<up> Done;
+            \<bottom> |
+          Cont (IntersectFound _) \<Rightarrow>
+            client_program ItemsUpdate \<kappa> \<psi> \<I>
+        ) |
+      ItemsUpdate \<Rightarrow>
+        \<up> Cont RequestNext;
+        \<down> M; (partial_case M of
+          Cont (RollForward i) \<Rightarrow>
+            client_program ph \<kappa> \<psi> (\<I> @ [i]) |
+          Cont (RollBackward p) \<Rightarrow>
+            client_program ph \<kappa> \<psi> (roll_back_to \<I> p \<psi>) |
+          Cont AwaitReply \<Rightarrow> \<comment> \<open>client is up to date\<close>
+            \<up> Done;
+            \<bottom>
+        )
+    )"
+
+definition index_of :: "'p \<Rightarrow> 'i list \<Rightarrow> ('i \<Rightarrow> 'p) \<Rightarrow> nat" where
+  [simp]: "index_of p \<I> \<psi> = (THE k. \<psi> (\<I> ! k) = p)"
+
+corec server_program where
+  "server_program rp mrb \<rho> \<psi> \<I> =
+    \<down> M; (partial_case M of
+      Done \<Rightarrow>
+        \<bottom> |
+      Cont (FindIntersect ps) \<Rightarrow> (
+        case \<rho> \<I> ps of
+          None \<Rightarrow>
+            \<up> Cont IntersectNotFound;
+            server_program rp mrb \<rho> \<psi> \<I> |
+          Some p \<Rightarrow>
+            \<up> Cont (IntersectFound p);
+            server_program (index_of p \<I> \<psi>) True \<rho> \<psi> \<I>
+      ) |
+      Cont RequestNext \<Rightarrow>
+        if mrb then
+          \<up> Cont (RollBackward (\<psi> (\<I> ! rp)));
+          server_program (Suc rp) False \<rho> \<psi> \<I>
+        else
+          if rp < length \<I> then
+            \<up> Cont (RollForward (\<I> ! rp));
+            server_program (Suc rp) mrb \<rho> \<psi> \<I>
+          else \<comment> \<open>client is up to date\<close>
+            \<up> Cont AwaitReply;
+            server_program rp mrb \<rho> \<psi> \<I>
```

Comment:
Since this is essentially a case distinction with 3 branches, we probably should adopt a different layout here, one that expresses such a 3-way case distinction at the expense of not being faithful to the actual term structure, which uses nesting of `if` expressions. I propose to place `if rp < length ℐ then` right after the preceding `else`, thus giving somewhat the impression that we use an `elif` construct here, and indent the following lines by one level less.

Reply:
Fixed in d2bebec6ebce866d3076b57d8ba64c4a0e9dc42f, please check.

Reply:
Hmm, I meant to place `else if k < length C then` on a single line.

Reply:
You meant only to place `else if k < length C then` on a single line and not indent the rest one level less?

Reply:
No, I meant *both*. This way it would look quite like a multi-way if expression in other languages, with `elif` replaced by `else if`:
```
if ⟨condition⟩ then
  […]
else if ⟨condition⟩ then
  […]
else
  […]
```

Reply:
Fixed in 2b8a8a3863ad1fc71d05cec3786480d6ba3916bb, please check.